benchmarking intial tests

amittal94 opened the pull request:
This is a possible approach to benchmarking the serialisation step. Please let me know what you think.

amittal94 committed:
benchmarking first version

amittal94 commented:
![image](https://github.com/user-attachments/assets/b23c99e9-6457-4d35-b2d4-86f8f5710c62)

stephen-jd-dann commented on tests/test_benchmarking.py:
```diff
@@ -0,0 +1,229 @@
+        "source_name": f"source_{random.randint(1, 1000)}",
+        "value": random.randint(1, 1000),
+        "timestamp_unix_ns": int(datetime.now().timestamp() * 1e9),
+        "alarm_status": random.randint(0, 22),
+        "alarm_severity": random.randint(0, 4),
```
These will usually be 0. In theory this will be slightly faster, so it might be worth addressing that.

Similar comments apply to other alarm fields.

amittal94 replied:
So just assigning it 0 is enough for benchmarking? I did think this would be unnecessary as it should have minimal impact to serialise say 15 as opposed to 0. Or do you mean something else?

stephen-jd-dann replied:
Sorry, I think it might be worth biasing the random number generator so it produces 0 more often. Ideally, either both these fields are 0 or both are non-zero. Or maybe you could benchmark those two cases separately?

amittal94 replied:
for now I have added a bias

stephen-jd-dann commented on tests/test_benchmarking.py:
```diff
@@ -0,0 +1,229 @@
+    return (), {
+        "source_name": f"source_{random.randint(1, 1000)}",
+        "unique_id": random.randint(1, 1000),
+        "data": np.random.randint(0, 1000, size=(1000, 1000), dtype=np.uint64),
```
This should be `uint8` or `uint16`.

stephen-jd-dann commented on tests/test_benchmarking.py:
```diff
@@ -0,0 +1,229 @@
+def generate_random_wa00_entry():
+    size = 1000
+    return (), {
+        "values_x_array": np.random.randint(0, 1000, size=size, dtype=np.uint64),
+        "values_y_array": np.random.randint(0, 1000, size=size, dtype=np.uint64),
```
These should be floating point I think? Although it can vary, `uint64` is very unlikely.

stephen-jd-dann reviewed (commented):
I like the general approach. I've jotted down a few notes on the data generation, but mainly looked at the benchmarking code proper.

- You're missing `pytest-benchmark` from the dependencies
- We shouldn't run benchmarks as part of the normal test suite. Good benchmarks take time. We can still run them in CI, but a normal call to `pytest` shouldn't run them (see what we've done for integration tests)
- You may not be using enough rounds for the faster benchmarks. A good benchmark should take at least 1 second
- Some warmup rounds might be useful. Newer versions of Python will need to warm up the JIT, but I don't know how long that would take
- It's unfortunate that you can't use a setup function with the automatic calibration. We may need to look for a better benchmarking library at some point

amittal94 commented:
I will need more explanation of the warm up rounds, do we really need them here? If so a manual approach adding it to the test vs pytests in built options may work better, as it seems to skip those flags based on its internal checks (I am not sure how to use it correctly).

stephen-jd-dann commented on benchmarks/test_serialisers.py:
```diff
@@ -10,19 +10,19 @@
         "timestamp_unix_ns": int(datetime.now().timestamp() * 1e9),
-        "alarm_status": random.randint(0, 22),
-        "alarm_severity": random.randint(0, 4),
+        "alarm_status": random.choice([0, random.randint(1, 22)]),
+        "alarm_severity": random.choice([0, random.randint(1, 4)]),
```
Ooh, that's a neat way of doing it.

I think the only realistic improvement would be benchmarking the two cases separately.

stephen-jd-dann commented on benchmarks/test_serialisers.py:
```diff
@@ -67,7 +67,7 @@ def generate_ntscalarany_dict():
             "precision": 2,
             "form": {
-                "index": random.randint(0, 6),
+                "index": random.choice([0, random.randint(1, 6)]),
```
Not really necessary here, but fine.

stephen-jd-dann commented on dev.py:
```diff
@@ -251,6 +251,16 @@ def cmd_test():
+
+    venv = require("venv")
+
+    venv.run_cmd("mypy", *mypy_check_paths())
```
I'd remove this line. No need to mypy again.

stephen-jd-dann commented on pyproject.toml:
```diff
@@ -43,3 +44,8 @@ exclude = [
+[tool.pytest.ini_options]
+testpaths = [
+    "tests",
+]
```
Nitpick: missing newline at end of file.

stephen-jd-dann reviewed (commented):
Okay, I take it back. Setting `rounds` to 10000 on every benchmark makes it too slow.

We can set it per benchmark, but running it a bit I'm not very happy with the statistics being produced. We can probably merge this, but I'd like to find a benchmarking system that handles setup functions better.

stephen-jd-dann commented:
Okay, so benchmarking is hard.

I've been fiddling around with this, trying to avoid use of pedantic mode by pre-generating data. Here's the idea for f142:

```python
def test_f142_benchmark(benchmark):
    random.seed(1234)
    entries = [generate_random_f142_entry()[1] for _ in range(100)]
    import itertools
    entries = itertools.cycle(entries)
    benchmark(lambda: serialise_f142(**next(entries)))
```

Note the use of `random.seed` to get consistent results.

However, as soon as I tried to extend this to `ntndarray` things started going wrong. I have two benchmarks:

```python
def test_ntndarray_benchmark_2(benchmark):
    random.seed(1234)
    entries = [generate_ntndarray_entry()[0][0] for _ in range(100)]
    import itertools
    entries = itertools.cycle(entries)
    entry = next(entries)
    benchmark.pedantic(lambda: serialise_data(entry), rounds=500)


def test_ntndarray_benchmark_3(benchmark):
    random.seed(1234)
    entries = [generate_ntndarray_entry()[0][0] for _ in range(100)]
    import itertools
    entries = itertools.cycle(entries)
    entry = next(entries)
    benchmark(lambda: serialise_data(entry))
```

When I run it with `--benchmark-max-time=5` (hint: use `-k` to limit which benchmarks are run) I get the following results:

```
------------------------------------------------------------------------------------ benchmark: 3 tests ------------------------------------------------------------------------------------
Name (time in ms)                 Min                Max              Mean            StdDev            Median               IQR            Outliers       OPS            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_ntndarray_benchmark_2     3.8618 (1.0)      14.5190 (1.0)      4.3764 (1.0)      0.6125 (1.0)      4.2236 (1.0)      0.4884 (1.0)         38;15  228.4990 (1.0)         500           1
test_ntndarray_benchmark       4.0920 (1.06)     16.1912 (1.12)     4.9531 (1.13)     1.3838 (2.26)     4.6493 (1.10)     0.6360 (1.30)          5;5  201.8952 (0.88)        100           1
test_ntndarray_benchmark_3     6.8330 (1.77)     16.7981 (1.16)     7.5915 (1.73)     0.6743 (1.10)     7.5373 (1.78)     0.6007 (1.23)        41;11  131.7259 (0.58)        361           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

Benchmark 3, using calibrated mode rather than pedantic mode, is **over 50% slower** than benchmark 2 for no obvious reason.

I've tried to reproduce this problem with some toy benchmarks, and I can't, so it's not a fundamental problem with `pytest-benchmark`. I also see significantly worse performance if using an inner function instead of a lambda.

Changing the image size has a big effect. I see no difference at 3e4, 6e4 seems to behave the same as 1e6, and at 2.5e5 it's even more extreme. I suspected a garbage collection issue, but disabling GC (`--benchmark-disable-gc`) does nothing. cProfile (`--benchmark-cprofile=cumtime`) suggests that the main difference is in `CreateNumpyVector`.

Frankly I'm out of ideas at this point. We can't use this as a benchmark as-is, because small changes that should be insignificant are having huge effects. Can you see if you can reproduce these issues?

[test_serialisers.py](https://github.com/user-attachments/files/19780784/test_serialisers.py.txt)

amittal94 commented:
I see, that is tricky, I will play around it more, and maybe consider different tooling. I have seen in the past that numpy vector creation (atleast in the case of converting from a python list) is a computationally expensive task, so I wonder if this is similar.

I am able to reproduce the results, although the discrepancy only shows up when I only run the test_ntndarray on their own, which I found strange within itself. Incidentally I was also produce the reversed results...

For reference at 2e5 I found benchmark_2 to be slower, at 2.5e5 benchmark_3 was slower and and 3e5 benchmark_2 was slower again. I will look for a different benchmarking approach.

So there seems to be a plateau till 1e5 where all 3 approaches perform fairly close, though 1 is slightly slower. For 1e5 and above 1 and 2 are faster than 3. I still don't see any clear reason.

amittal94 committed:
alternate approach to benchmarking

amittal94 commented:
maybe this alternative approach may work better? Data generation is explicitly separate from the benchmark, so it should really not interfere. I also kept the shape the same for adar as in ntndarray, to avoid reshape interference.

stephen-jd-dann commented on benchmarks/benchmark_serialisers.py:
```diff
@@ -0,0 +1,267 @@
+
+    for i in range(count):
+        start = time.perf_counter_ns()
+        benchmark_fn(**entry)
```
This is just picking the last entry instead of looping over them.

amittal94 replied:
Oops, did not spot this, I will fix it.

amittal94 replied:
Actually, now that I think about it, is it perhaps better to just use a single value for testing?

stephen-jd-dann replied:
Well, maybe.

The thing you want to avoid is that the value you're testing isn't representative of the real data. It might run faster or slower in real life, and optimisations might only apply to the data you use for benchmarking.

But I don't think that's a problem here. The things we're randomising don't really affect the speed, and the things that probably affect the speed aren't randomised.

So it's probably simplest to just use a single value, but if it's randomly generated you should use a fixed seed so it's the same value every time.

stephen-jd-dann reviewed (commented):
We're in serious danger of going down a rabbit hole here. I'm going to muck about with your custom benchmark code and see how stable it is. I think we should try to get `pytest-bench` working if possible.

stephen-jd-dann commented:
Your custom benchmark code makes it fairly easy to loop over the array size and do lots of benchmarks. Helpful hints when doing this:
1. Randomise (`random.shuffle`) the order in which you iterate to reduce systematic errors
2. Go for a fixed target time instead of a fixed number of iterations
3. Run the whole thing at least three times, so you can plot the median, min, and max

Here's what I got:
![Untitled-1](https://github.com/user-attachments/assets/e3e0c5c5-6ed7-449b-8e77-9782e2c0b83f)

That jump at 1e6 elements is very clear. Here's a zoomed-in view:

![Untitled](https://github.com/user-attachments/assets/c1487c08-9c59-438f-b92d-162a42ef2f04)

You can sort-of see similar features around 5e5 and 2.5e5. I suspect this has something to do with memory allocation. A common strategy when expanding a dynamic array (like the byte buffer being serialised into) is to double the size each time it fills up. These cut-off points are probably powers of 2.

That's not what's happening here, because I did a test where I increased the starting size of the buffer to a number that should be big enough and the edge at 1e6 or 2^20 or wherever it is was still there.

Anyway, some parts of this curve seem to be well behaved, so maybe we can do all our benchmarking there and we can go back to using `pytest-bench` and everything will be fine? It seemed well-behaved when I tried it with `size=700000` (that's 7e5). But we'll need a comment explaining why the exact size value is important.

stephen-jd-dann commented:
In case you're interested, benchmarking the following function is enough to reproduce the edge at about 1.05e6:

```python
def test_func(data: dt.PVData):
    v = data.data.value
    buffer = bytearray(1024 * 1024 * 8 * 2)
    array_bytes = v.tobytes()
    copy_len = len(array_bytes)
    offset = 100
    buffer[offset:offset+copy_len] = array_bytes
    return buffer
```

Recall that the array is made of uint64, so the actual array size is 8 MiB.

I've tried it on a CPU with bigger caches and the difference is bigger if anything, so I don't think it's the CPU cache.

amittal94 commented:
I did similar testing, with fairly similar results, with most noticeable jumps being at 0.5e5 and 1e6. Stable points seem to be around 3.5e5 and 8e5. I think perhaps it will be best to just simply use one version of the pytest-benchmark tests and keep the simpler version as an extra approach.

stephen-jd-dann commented:
Yes, I agree, with the caveat that we should make sure we're testing at a stable point and leave a comment explaining why.

amittal94 committed:
refactor and simplify tests

- separate tests by data type
- shift to non pedantic tests
- remove randomness in favor of single entry

stephen-jd-dann commented on benchmarks/test_images.py:
```diff
@@ -0,0 +1,170 @@
+from epac.flatbuffers import data_types as dt
+
+
+def generate_adar_entry(size):
```
Should this be `generate_adar_dict` for consistency?

stephen-jd-dann commented on benchmarks/test_images.py:
```diff
@@ -0,0 +1,170 @@
+
+@pytest.mark.parametrize(
+    "adar_entry",
+    [generate_adar_entry(i * 300) for i in range(1, 5)],
```
I'd parameterise it with just the size and do the generation inside the test (but outside the benchmark). Means you don't need the `ids=` I think.

stephen-jd-dann replied:
Maybe a wider range of sizes, which could be kept in a global variable to sync them between the two tests?

10, 30, 100, 300, 1000 might be a better range, though we should try to avoid the awkward sizes.

stephen-jd-dann commented on benchmarks/test_images.py:
```diff
@@ -0,0 +1,170 @@
+    benchmark(lambda: serialise_data(ntndarray_entry))
+
+
+# pedantic based options
+# def test_ntndarray_benchmark_p(benchmark):
+#     benchmark.pedantic(lambda: serialise_data(ntndarray_entry), rounds=100)
+
+# def test_ntndarray_benchmark_p3(benchmark):
+#     def foo():
+#         serialise_data(ntndarray_entry)
+
+#     benchmark.pedantic(foo, rounds=100)
```
Can you remove these or provide more detail about what's going on?

stephen-jd-dann commented on benchmarks/test_scalars.py:
```diff
@@ -0,0 +1,74 @@
+from epac.flatbuffers import data_types as dt
+
+
+def generate_f142_entry():
```
Again, naming might be better as `_dict`

stephen-jd-dann reviewed (changes requested):
There's a stray comment that needs to be removed.

Other than that it looks good, although you could try a wider range of sizes for the images. A geometric series might be nice there.
